dowhy/gcm/unit_change.py

```diff
@@ -99,6 +100,60 @@ def unit_change_linear(
     return contribution_df
 
 
+def unit_change_nonlinear_input_only(
+    mechanism: PredictionModel,
+    background_df: pd.DataFrame,
+    foreground_df: pd.DataFrame,
+    input_column_names: List[str],
+    shapley_config: Optional[ShapleyConfig] = None,
+) -> pd.DataFrame:
+    """
+    Calculates the contributions of each input to the change in the output values of a non-linear deterministic mechanism.
+    The technical method is a modification of the attribution method described in the following research paper, without mechanism as a player:
+    Kailash Budhathoki, George Michailidis, Dominik Janzing. *Explaining the root causes of unit-level changes*. arXiv, 2022.
+
+    :param mechanism: The mechanism.
+    :param background_df: The background data.
+    :param foreground_df: The foreground data.
+    :param input_column_names: The names of the input (features) columns in both dataframes.
+    :param shapley_config: The configuration for calculating Shapley values.
+    :return: A pandas dataframe with attributions to each cause for the change in each output row of provided dataframes.
+    """
+    _check_if_input_columns_exist(background_df, foreground_df, input_column_names)
+
+    def payoff(player_indicator: List[int]) -> np.ndarray:
+        input_arrays = []
+        for i, is_player_active in enumerate(player_indicator):
+            selected_df = foreground_df if is_player_active else background_df
+            input_arrays.append(selected_df[input_column_names[i]].to_numpy())
+        return mechanism.predict(np.column_stack(input_arrays)).flatten()
+
+    contributions = estimate_shapley_values(payoff, len(input_column_names), shapley_config)
+    return pd.DataFrame(contributions, columns=input_column_names)
+
+
+def unit_change_linear_input_only(
```

Comment:
What about using ```if isinstance(mechanism, LinearPredictionModel)``` and then have only one method ```unit_change_input_only```? Same applies to the existing method.

To further simplify it (i.e., avoid having multiple different methods), you could also consider adding a parameter ```analyze_inputs_only``` to the other existing method. Then, there would only be one API to call that can be configured and only requires one additional parameter. Or are these two approaches fundamentally different?

Reply:
> Under a single method, a user still has to select an appropriate wrapper (SklearnLinearRegressionModel or SklearnRegressionModel) for sklearn models, and pass its object

That's right, one need to check if the model ```has_coefficients``` actually, since this is what you need to the linear solution. I see the concern here. However, it would probably simplify the usage seeing that, if we have ```has_coefficients```, then the ```_linear``` version represents the closed-form solution that avoids the Shapley estimation. In that sense, there is no reason to use the ```_nonlinear``` version with a linear model.

I guess the question here is: Do we want to go a route where the methods automatically optimize the computation based on the model type (which requires ugly things like ```if isinstance(x, y)``` or should this be the responsibility of the user?

Reply:
Force pushed a single helper API for all those four APIs.